File: common/rpc/RpcEventHandlerBase.cpp


#include "RpcEventHandlerBase.h"

using namespace zonda::common;
int RpcEventHandlerBase::handle_rpc_event(const IEvent* event)
{
    RpcEvent* e = (RpcEvent*)event;
    int r=0;
    RpcHead* rpc_head = e->get_rpc_head();
    if (e->is_req())
    {
        r = handle_rpc_req(e->get_op_type(),
                e->get_msg(),
                rpc_head);
    }
    else
    {
        r = handle_rpc_res(
                e->get_op_type(),
                rpc_head->res_code,
                e->get_msg(),
                rpc_head->trans_param,
                rpc_head->transaction_id,
                e->get_req());
    }
    
    delete e;
    return r;
}

File: common_java/src/main/java/com/alibaba/platform/zonda/common/rpc/RpcEventHandlerBase.java
package com.alibaba.platform.zonda.common.rpc;

import com.alibaba.platform.zonda.common.seda.*;

public abstract class RpcEventHandlerBase implements IEventHandler
{
    //! When an rpc request arrives, the method is called
    public void handleRpcReq(
        int opType,
        IMsg msg,
        RpcHead rpcHead)
    {
    }
     
    public abstract void handleRpcRes(
        int opType,
        int resCode,
        IMsg msg,
        long transParam,
        long transactionId,
        IMsg originalReq);
    
    public void handleEvent(IEvent event)
    {
        if (event.getType() != EventType.RPC_EVENT)
        {
            throw new RuntimeException("Unknow EventType:" + event.getType());
        }
        RpcEvent e = (RpcEvent)event;
        RpcHead rpcHead = e.getRpcHead();
        if (e.isReq())
        {
            handleRpcReq(e.getOpType(),
                    e.getMsg(),
                    rpcHead);
        }
        else
        {
            
            handleRpcRes(
                e.getOpType(),
                rpcHead.resCode.get(),
                e.getMsg(),
                rpcHead.transParam.get(),
                rpcHead.transactionId.get(),
                e.getReq());
        }
    
    }    
}